RFCs/0010-Source-Information-For-Discovered-Tests.md

```diff
@@ -0,0 +1,69 @@
+# 0010 Source Information for discovered tests
+
+## Summary
+This note outlines the proposed change in behavior of source information collection for test methods discovered by test adapters.
+This change is applicable for test discovery from:
+1. Visual Studio Test Explorer (IDE) for managed test projects which are supported by .NET Compiler Platform (Roslyn)
+2. Vstest.console.exe runner (CLI) 
+
+## Motivation
+a. Performance improvement in IDE scenario because of two factors:
+* Use of better performant Roslyn APIs instead of reflection based approach on PDBs
+* Reduced test method payload transfer from adapter to Test Explorer
+
+b. Performance improvement in CLI scenario:
+* Source information is not required in CLI scenario and can be skipped entirely
+
+## Overview of changes
+Visual Studio Test Explorer needs source information for discovered tests to power following 3 scenarios:
+1. Navigation to test method
+2. Test method execution from code editor context menu
+3. Code lens markers
+
+At present, for managed (C# and VB) test projects, this information is obtained from PDBs using reflection based APIs. While this approach works, it has perf impact.
+Performance can be improved by using Roslyn APIs which gather this information during compile time.
+
+During discovery, Visual Studio Test Explorer groups the test containers on basis of their target framework (E.g.: NET46, NetCore etc.). 
+For each group a separate discovery request is sent to the adapters. 
+If all test containers in a group are supported by Roslyn, then for that discovery request, Visual Studio Test Explorer sets **'CollectSourceInformation'** flag in 
+runsettings with value 'false' and passes it to the test adapters, else it is set to 'true'.
+Test adapters can rely on this flag and skip source information collection for discovered tests, if the flag is set to 'false'. In such cases, Visual Studio Test Explorer
+will determine source information for the discovered tests.
+
+In case of CLI, source information is not required. Hence this flag is set to 'false' in the runsettings.
+
+### A sample runsettings file with this flag:
+```xml
+<RunSettings>
+    <RunConfiguration>
+      <CollectSourceInformation>False</CollectSourceInformation>
+    </RunConfiguration>
+</RunSettings>
+```
+
+Following are the perf numbers obtained for this flag in IDE scenario for MSTest and XUnit test adapters on TPv2 for a test project having 25K tests:
+
+| Adapter         | Without Changes = A (Sec) | With Changes = B (Sec) | Improvement, C = A - B (Sec) | Improvement (%) = C * 100 / A |
```

Comment:
Can we validate the perf benefits in nunit too? /cc: @jnm2 @rprouse @OsirisTerje